src/Data/Arena.hs

```diff
@@ -15,74 +17,169 @@ module Data.Arena (
+    marena <- atomicModifyMutVar' arenas $ \case
+        []     -> ([], Nothing)
+        (x:xs) -> (xs, Just x)
 
-closeArena :: PrimMonad m => Arena (PrimState m) -> m ()
-#ifdef NO_IGNORE_ASSERTS
-closeArena (Arena mvar) = do
-    -- scramble the allocated bytearrays,
-    -- they shouldn't be in use anymore!
-    mbas <- readMutVar mvar
-    forM_ mbas $ \mba -> do
-        size <- getSizeofMutableByteArray mba
-        setByteArray mba 0 size (0x77 :: Word8)
+    case marena of
+        Just arena -> return arena
+        Nothing -> do
+            curr <- newBlock >>= newMVar
+            free <- newMutVar []
+            full <- newMutVar []
+            return Arena {..}
+
+closeArena :: PrimMonad m => ArenaManager (PrimState m) -> Arena (PrimState m) -> m ()
+closeArena (ArenaManager arenas) arena = do
+    scrambleArena arena
+
+    -- reset the arena to clear state
+    resetArena arena
+
+    atomicModifyMutVar' arenas $ \xs -> (arena : xs, ())
+
+
+
+scrambleArena :: PrimMonad m => Arena (PrimState m) -> m ()
+#ifndef NO_IGNORE_ASSERTS
+scrambleArena _ = return ()
 #else
-closeArena _ = pure ()
+scrambleArena Arena {..} = do
+    readMVar curr >>= scrambleBlock
+    readMutVar full >>= mapM_ scrambleBlock
+    readMutVar free >>= mapM_ scrambleBlock
+
+scrambleBlock :: PrimMonad m => Block (PrimState m) -> m ()
+scrambleBlock (Block _ mba) = do
+    size <- getSizeofMutableByteArray mba
+    setByteArray mba 0 size (0x77 :: Word8)
 #endif
 
--- | Create unmanaged arena
+-- | Reset arena, i.e. return used blocks to free list.
+resetArena :: PrimMonad m => Arena (PrimState m) -> m ()
+resetArena Arena {..} = do
+    Block off mba <- takeMVar curr
+
+    -- reset current block
+    writePrimVar off 0
+
+    -- move full block to free blocks.
+    -- block's offset will be reset in 'newBlockWithFree'
+    full' <- atomicModifyMutVar' full $ \xs -> ([], xs)
+    atomicModifyMutVar' free $ \xs -> (full' <> xs, ())
+
+    putMVar curr $! Block off mba
+
+-- | Create unmanaged arena.
 --
 -- Never use this in non-tests code.
 withUnmanagedArena :: PrimMonad m => (Arena (PrimState m) -> m a) -> m a
 withUnmanagedArena k = do
     mgr <- newArenaManager
     withArena mgr k
 
+-- | Allocate a slice of mutable byte array from the arena.
 allocateFromArena :: PrimMonad m => Arena (PrimState m)-> Size -> Alignment -> m (Offset, MutableByteArray (PrimState m))
-allocateFromArena (Arena mvar) !size !alignment = do
-    mba <- newAlignedPinnedByteArray size alignment
-    atomicModifyMutVar' mvar $ \mbas -> (mba : mbas, ())
-    return (0, mba)
+allocateFromArena !arena !size !alignment =
+    assert (popCount alignment == 1) $ -- powers of 2
+    assert (size <= blockSize) $ -- not to large allocations
+    allocateFromArena' arena size alignment
+
+-- TODO!? this is not async exception safe
+allocateFromArena' :: PrimMonad m => Arena (PrimState m)-> Size -> Alignment -> m (Offset, MutableByteArray (PrimState m))
```

Comment:
Might not be too bad, but tell me if I misunderstood something. :thinking: 

If it fails before putting back the MVar, the arena becomes unusable. Future calls will deadlock, but at least not silently corrupt things.
And AFAIU the way we use the arena for lookups, that should be fine, since the exception will bubble up (without a call to `closeArena`) and might only be caught outside of `lookup`. Retrying there will give you a new arena. The unusable one should at some point get garbage collected? So it's a bit wasteful, but only in exceptional situations.

Maybe that's good enough and we just want to document it?

Reply:
It really depends whether the code is run in IO or ST. Your analysis is probably true for ST usecases, for IO it's not.

Reply:
Ah, interesting. What might go wrong in IO?

Reply:
there is `bracket`, so one could actually recover. (This is why I don't like having `newArena/closeArena` exported, I see that as a deficit in benchmarking library, that it doesn't support `with*` pattern resource allocation).

Reply:
Do you mean using `bracket _ closeArena` in `withArena`? This would require us to also use `bracket`/`withMVar` within `allocateFromArena`, too, right (if we don't put back the MVar, `closeArena` will hang)? But then we have to be careful with the state of the other vars. E.g. we put back `curr'`, but we had already added it to `full` as well. Or we got something from `free`, but end up just dropping it. Should be possible, but not trivial.

Reply:
@mheinzel I do think that *not* using `bracket` in `withArena` makes stuff simpler. In case of exception an invalid state arena gets GCd. Current interface is suboptimal.

Getting `allocateFromArena` exception safe might not be easy (I'd really like to use TVars, but those are not possible in ST - or then one needs another abstraction, TVars for IO and ordinary vars for ST).

Reply:
Okay, thanks. Even with an `ST` interface, someone could do something like `bracket (stToIO newArena) (stToIO . closeArena)`. And it indeed seems complicated to make `allocateFromArena` exception safe. So maybe a reasonable way forward would be to make sure non-test code always used `withArena` and document the situation:

- export `newArena` and `closeArena` under the "Test helpers" section, document that the arena won't be reused, so non-test code should use `withArena`, maybe rename them to make it really clear
- document `allocateFromArena` to say that async exceptions will leave the arena unusable, so it is suggested to let them bubble up to escape `withArena`

Reply:
For `closeArena`, it's only used in the cleanup of the lookup micro benchmark where it could even lead to a deadlock I think?

```haskell
          perRunEnvWithCleanup
            ( newArena arenaManager >>= \arena ->
              stToIO (prepLookups arena blooms indexes kopsFiles ks) >>= \(rkixs, ioops) ->
              FS.submitIO hasBlockIO ioops >>= \ioress ->
              pure (rkixs, ioops, ioress, arena)
            )
            (\(_, _, _, arena) -> closeArena arena) $ \ ~(rkixs, ioops, ioress, _) -> do
              !_ <- intraPageLookups resolveV rs ks rkixs ioops ioress
              pure ()
```

I guess we want to reuse the arena in the happy path here. But we could close the arena below `intraPageLookups`, so it's skipped on exception.

Reply:
I think there is value in exposing `newArena` and `closeArena`, even if it is only for unsafe uses. We can prefix them with `unsafe` instead.

Regarding `closeArena` hanging, maybe we should then have an `unsafeTryCloseArena`

Reply:
So should I change something here, or can we merge this?